tests/StratigraphyControllerTest.cs

```diff
@@ -0,0 +1,153 @@
+        }
+        finally
+        {
+            // Delete stratigraphy copy
+            if (copiedStratigraphy != null)
+            {
+                context.Layers.RemoveRange(copiedStratigraphy.Layers);
+                context.LithologicalDescriptions.RemoveRange(copiedStratigraphy.LithologicalDescriptions);
+                context.FaciesDescriptions.RemoveRange(copiedStratigraphy.FaciesDescriptions);
+                context.Stratigraphies.Remove(copiedStratigraphy);
+                context.SaveChanges();
+            }
+        }
+    }
+
+    private Stratigraphy GetStratigraphy(int id)
+    {
+        return context.Stratigraphies
+            .Include(s => s.CreatedBy)
+            .Include(s => s.UpdatedBy)
+            .Include(s => s.Kind)
+            .Include(s => s.FillCasing)
+            .Include(s => s.Layers).ThenInclude(l => l.LayerCodelists)
+            .Include(s => s.LithologicalDescriptions)
+            .Include(s => s.FaciesDescriptions)
+            .Single(s => s.Id == id);
+    }
+
+    [TestMethod]
+    public async Task CopyInvalidStratigraphyId()
+    {
+        var result = await controller.CopyAsync(0).ConfigureAwait(false);
+        Assert.IsInstanceOfType(result.Result, typeof(NotFoundResult));
+    }
+
+    [TestMethod]
+    public async Task CopyWithUnknownUser()
+    {
+        controller.HttpContext.SetClaimsPrincipal("NON-EXISTENT-NAME", PolicyNames.Admin);
+        var result = await controller.CopyAsync(StratigraphyId).ConfigureAwait(false);
+        Assert.IsInstanceOfType(result.Result, typeof(UnauthorizedResult));
+    }
+
+    [TestMethod]
+    public async Task CopyWithUserNotSet()
+    {
+        controller.ControllerContext.HttpContext.User = null;
+        await Assert.ThrowsExceptionAsync<InvalidOperationException>(async () =>
+        {
+            await controller.CopyAsync(StratigraphyId).ConfigureAwait(false);
+        });
+    }
+
+    [TestMethod]
+    public async Task CopyWithNonAdminUser()
+    {
+        controller.HttpContext.SetClaimsPrincipal("editor", PolicyNames.Viewer);
+        var result = await controller.CopyAsync(StratigraphyId).ConfigureAwait(false);
+        Assert.IsInstanceOfType(result.Result, typeof(OkObjectResult));
+
+        // delete stratigraphy copy
+        var copiedStratigraphyId = ((OkObjectResult?)result.Result)?.Value;
+        Assert.IsNotNull(copiedStratigraphyId);
+        Assert.IsInstanceOfType(copiedStratigraphyId, typeof(int));
+        var copiedStratigraphy = context.Stratigraphies.Single(s => s.Id == (int)copiedStratigraphyId);
+        context.Layers.RemoveRange(copiedStratigraphy.Layers);
+        context.LithologicalDescriptions.RemoveRange(copiedStratigraphy.LithologicalDescriptions);
+        context.FaciesDescriptions.RemoveRange(copiedStratigraphy.FaciesDescriptions);
+        context.Stratigraphies.Remove(copiedStratigraphy);
+        context.SaveChanges();
```

Comment:
REQ: Das Abräumen der Daten in einen `finally`-Block packen, damit das in jedem Fall ausgeführt wird. Vermutlich lässt sich der Code aus dem Test `Copy()` auslagern und wiederverwenden.